From b153747cd2a849d1236c3431e140fb3b1c251ebb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Pierret?= <frederic.epitre@orange.fr>
Date: Sun, 1 Oct 2017 16:15:27 +0200
Subject: [PATCH 1/3] mce: hide EBUSY initialization error on Xen
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

In case of Xen, the device is already registered by xen mcelog (in
xen_late_init_mcelog), so fail here is expected. Note that
mcheck_init_device call is still expected to initialize mce_device. Comment
from threshold_init_device explaining the situation:

/*
 * there are 3 funcs which need to be _initcalled in a logic sequence:
 * 1. xen_late_init_mcelog
 * 2. mcheck_init_device
 * 3. threshold_init_device
 *
 * xen_late_init_mcelog must register xen_mce_chrdev_device before
 * native mce_chrdev_device registration if running under xen platform;
 *
 * mcheck_init_device should be inited before threshold_init_device to
 * initialize mce_device, otherwise a NULL ptr dereference will cause panic.
 *
 * so we use following _initcalls
 * 1. device_initcall(xen_late_init_mcelog);
 * 2. device_initcall_sync(mcheck_init_device);
 * 3. late_initcall(threshold_init_device);
 *
 * when running under xen, the initcall order is 1,2,3;
 * on baremetal, we skip 1 and we do only 2 and 3.
 */

Source: Marek Marczykowski-Górecki <marmarek@invisiblethingslab.co>
Signed-off-by: Frédéric Pierret <frederic.epitre@orange.fr>
---
 arch/x86/kernel/cpu/mcheck/mce.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/arch/x86/kernel/cpu/mcheck/mce.c b/arch/x86/kernel/cpu/mcheck/mce.c
index 3b413065..55b7d1d 100644
--- a/arch/x86/kernel/cpu/mcheck/mce.c
+++ b/arch/x86/kernel/cpu/mcheck/mce.c
@@ -53,6 +53,10 @@
 #include <asm/reboot.h>
 #include <asm/set_memory.h>
 
+#ifdef CONFIG_XEN_MCE_LOG
+#include <xen/xen.h>
+#endif
+
 #include "mce-internal.h"
 
 static DEFINE_MUTEX(mce_log_mutex);
@@ -2355,6 +2359,11 @@ static __init int mcheck_init_device(void)
 	free_cpumask_var(mce_device_initialized);
 
 err_out:
+#ifdef CONFIG_XEN_MCE_LOG
+	/* in case of Xen, the character device was already registered, so do not
+	 * treat this as an error */
+	if (!xen_initial_domain() || err != -EBUSY)
+#endif
 	pr_err("Unable to init MCE device (rc: %d)\n", err);
 
 	return err;
-- 
2.9.5

